vowpalwabbit/options.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <typeinfo>
+#include <memory>
+
+namespace VW {
+  namespace config {
+
+  struct base_option {
+    base_option(std::string name, size_t type_hash)
+      : m_name(name), m_type_hash(type_hash)
+    {}
+
+    std::string m_name;
+    size_t m_type_hash;
+    std::string m_help = "";
+    std::string m_short_name = "";
+    bool m_keep = false;
+
+    virtual ~base_option() {}
+  };
+
+  template<typename T>
+  struct typed_option : base_option {
+    typed_option(std::string name, T& location)
+      : base_option(name, typeid(T).hash_code()), m_location{ location } {}
+
+    static size_t type_hash() {
+      return typeid(T).hash_code();
+    }
+
+    typed_option& default_value(T value) {
+      m_default_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    bool default_value_supplied() {
+      return m_default_value.get() != nullptr;
+    }
+
+    T default_value() {
+      return m_default_value ? *m_default_value : T();
+    }
+
+    typed_option& short_name(std::string short_name) {
+      m_short_name = short_name;
+      return *this;
+    }
+
+    typed_option& help(std::string help) {
+      m_help = help; return *this;
+    }
+
+    typed_option& keep(bool keep = true) {
+      m_keep = keep; return *this;
+    }
+
+    bool value_supplied() {
+      return m_value.get() != nullptr;
+    }
+
+    typed_option& value(T value) {
+      m_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    T value(){
+      return m_value ? *m_value : T();
+    }
+
+    T& m_location;
+
+  private:
+    // Would prefer to use std::optional (C++17) here but we are targeting C++11
+    std::shared_ptr<T> m_value{ nullptr };
+    std::shared_ptr<T> m_default_value{ nullptr };
+  };
+
+  template<typename T>
+  typed_option<T> make_typed_option(std::string name, T& location) {
+    return typed_option<T>(name, location);
+  }
+
+  struct option_group_definition {
+    option_group_definition(std::string name)
+      : m_name(name)
+    {}
+
+    template<typename T>
+    void add(typed_option<T> op) {
+      m_options.push_back(std::make_shared<typed_option<T>>(op));
+    }
+
+    template<typename T>
+    option_group_definition& operator()(typed_option<T> op) {
+      add(op);
+      return *this;
+    }
+
+    std::string m_name;
+    std::vector<std::shared_ptr<base_option>> m_options;

```

Comment:
>shared_ptr [](start = 21, length = 10)

why vector<shared_ptr> and not just vector<option>?

Reply:
Options may be retrieved from the class and shouldn't be copied, but their lifetime needs to ensured. It may be possible to use references instead but I didn't go that far

Reply:
1) what's the reason copies are not allowed
2) why not prevent copies using a private copy constructor?
3) since the vector is exposed, why not let the vector handle returning the reference?

---
In reply to: [246873582](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1706#discussion_r246873582) [](ancestors = 246873582)

Reply:
options_i returns a shared_ptr instead of a reference.  Is that the issue?


---
In reply to: [246889589](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1706#discussion_r246889589) [](ancestors = 246889589,246873582)

Reply:
They are created as part of the option_group_definition, then passed in and ownership is taken over by the options class.

It's a tricky set of lifetimes. But I may be able to make it so that only rvalues are accepted when adding to the option_group_definition, and then the values stored in the option_group_definition are processed they are then moved to the options object.

As you can see the typed_option object gets moved around a fair bit, so I opted for using a pointer to help ensure it isn't copied and one value is kept alive.